### PR TITLE
[User profile] : Extending standard data with custom template

### PR DESCRIPTION
**Pull request by NicolasEYSSERIC**
A custom template can be associated to directory and users profile. Only once template for the moment.
This template is displayed in front office (to user himself in edition and other users) and back office too.
This template is available in directory behind the advanced search. Do not forget to make this template "searchable" and some of its fields...

At this occasion, several improvements :
-  user page in back-office has been redesigned,
- admin is able to delete user's avatar.

Do not forget to check out PR on Silverpeas-Components...

### CODE REVIEW

File: core-war/src/main/java/org/silverpeas/web/socialnetwork/myprofil/control/MyProfilSessionController.java [hunk 1]
```diff
@@ -229,4 +248,22 @@ public boolean updatablePropertyExists() {
...
\ No newline at end of file
+
+  private void processDataOfExtraTemplate(HttpRequest request) {
+    PublicationTemplateManager templateManager = PublicationTemplateManager.getInstance();
+    PublicationTemplate template = templateManager.getDirectoryTemplate();
+    if (template != null) {
+      try {
+        PagesContext context = getTemplateContext();
+        templateManager.saveData(template.getFileName(), context, request.getFileItems());
+      } catch (Exception e) {
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/socialnetwork/myprofil/control/MyProfilSessionController.java]: MAJOR Either log or rethrow this exception. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1166)

File: core-war/src/main/java/org/silverpeas/web/socialnetwork/myprofil/control/MyProfilSessionController.java [hunk 2]
```diff
@@ -229,4 +248,22 @@ public boolean updatablePropertyExists() {
...
+        MessageNotifier.addError("Les données du formulaire n'ont pas été enregistrées !")
+            .setDisplayLiveTime(10000);
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/socialnetwork/myprofil/control/MyProfilSessionController.java]: MINOR Assign this magic number 10000 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)

File: core-war/src/main/java/org/silverpeas/web/socialnetwork/myprofil/control/MyProfilSessionController.java [hunk 3]
```diff
@@ -141,12 +147,25 @@ public void modifyUser(String idUser, String userLastName, String userFirstName,
...
-            (isAdmin() && theModifiedUser.isPropertyUpdatableByAdmin(property.getKey()))) {
-          theModifiedUser.setValue(property.getKey(), property.getValue());
+      Enumeration<String> parameters = request.getParameterNames();
+      String parameterName;
+      String property;
+      while (parameters.hasMoreElements()) {
+        parameterName = parameters.nextElement();
+        if (parameterName.startsWith("prop_")) {
+          // remove s"prop_"
+          property = parameterName.substring(5, parameterName.length());
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/socialnetwork/myprofil/control/MyProfilSessionController.java]: MINOR Assign this magic number 5 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)

File: core-library/src/main/java/org/silverpeas/core/admin/user/UserIndexation.java [hunk 4]
```diff
@@ -25,6 +25,10 @@
 
 import org.silverpeas.core.admin.user.model.UserDetail;
```
Comment (silveruser) [inline core-library/src/main/java/org/silverpeas/core/admin/user/UserIndexation.java]: MINOR Remove this unused import 'org.silverpeas.core.admin.user.model.UserDetail'. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core-library/src/main/java/org/silverpeas/core/admin/user/UserIndexation.java [hunk 5]
```diff
@@ -25,6 +25,10 @@
...
 import org.silverpeas.core.admin.user.model.UserFull;
+import org.silverpeas.core.contribution.content.form.RecordSet;
```
Comment (silveruser) [inline core-library/src/main/java/org/silverpeas/core/admin/user/UserIndexation.java]: MINOR Remove this unused import 'org.silverpeas.core.contribution.content.form.RecordSet'. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core-library/src/main/java/org/silverpeas/core/admin/user/UserIndexation.java [hunk 6]
```diff
@@ -25,6 +25,10 @@
...
+import org.silverpeas.core.contribution.template.publication.PublicationTemplate;
+import org.silverpeas.core.contribution.template.publication.PublicationTemplateManager;
+import org.silverpeas.core.exception.SilverpeasRuntimeException;
```
Comment (silveruser) [inline core-library/src/main/java/org/silverpeas/core/admin/user/UserIndexation.java]: MINOR Remove this unused import 'org.silverpeas.core.exception.SilverpeasRuntimeException'. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core-library/src/main/java/org/silverpeas/core/contribution/template/publication/PublicationTemplateManager.java [hunk 7]
```diff
@@ -510,4 +527,136 @@ protected void registerForRenewingContentCipher() {
...
+
+  private PublicationTemplate getTemplateAndRegisterIt(String xmlFormName, PagesContext context)
+      throws SilverpeasException {
+    String shortName = getShortName(xmlFormName);
+    String externalId = context.getComponentId() + ":" + shortName;
+    PublicationTemplate pub;
+    try {
+      // récupération des données du formulaire (via le DataRecord)
+      pub = getPublicationTemplate(externalId);
+    } catch (Exception e) {
```
Comment (silveruser) [inline core-library/src/main/java/org/silverpeas/core/contribution/template/publication/PublicationTemplateManager.java]: MAJOR Either log or rethrow this exception. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1166)

File: core-library/src/main/java/org/silverpeas/core/contribution/template/publication/PublicationTemplateManager.java [hunk 8]
```diff
@@ -510,4 +527,136 @@ protected void registerForRenewingContentCipher() {
...
+      PublicationTemplate usedTemplate = getPublicationTemplate(componentId + ":" + shortName);
+      RecordSet set = usedTemplate.getRecordSet();
+      set.indexRecord(userId, shortName, indexEntry);
+    } catch (Exception e) {
+      SilverLogger.getLogger(this).error(e);
+    }
+  }
+
+  public PublicationTemplate getDirectoryTemplate() {
+    List<PublicationTemplate> templates = getDirectoryTemplates();
```
Comment (silveruser) [inline core-library/src/main/java/org/silverpeas/core/contribution/template/publication/PublicationTemplateManager.java]: MAJOR Rename "templates" which hides the field declared at line 89. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AHiddenFieldCheck)

File: core-war/src/main/java/org/silverpeas/web/jobdomain/control/JobDomainPeasSessionController.java [hunk 9]
```diff
@@ -2137,4 +2135,27 @@ private UserDetail getUser(String userId, UserDetail[] users) {
...
-}
+
+  private void processDataOfExtraTemplate(String userId, HttpRequest request) {
+    PublicationTemplateManager templateManager = PublicationTemplateManager.getInstance();
+    PublicationTemplate template = templateManager.getDirectoryTemplate();
+    if (template != null) {
+      try {
+        PagesContext context = getTemplateContext(userId);
+        templateManager.saveData(template.getFileName(), context, request.getFileItems());
+      } catch (Exception e) {
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/jobdomain/control/JobDomainPeasSessionController.java]: MAJOR Either log or rethrow this exception. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1166)

File: core-war/src/main/java/org/silverpeas/web/jobdomain/control/JobDomainPeasSessionController.java [hunk 10]
```diff
@@ -2137,4 +2135,27 @@ private UserDetail getUser(String userId, UserDetail[] users) {
...
+        MessageNotifier.addError("Les données du formulaire n'ont pas été enregistrées !")
+            .setDisplayLiveTime(10000);
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/jobdomain/control/JobDomainPeasSessionController.java]: MINOR Assign this magic number 10000 to a well-named constant, and use the constant instead. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS109)

File: core-war/src/main/java/org/silverpeas/web/jobdomain/control/JobDomainPeasSessionController.java [hunk 11]
```diff
@@ -44,6 +44,10 @@
 import org.silverpeas.core.admin.user.model.GroupProfileInst;
 import org.silverpeas.core.admin.user.model.UserDetail;
 import org.silverpeas.core.admin.user.model.UserFull;
+import org.silverpeas.core.contribution.content.form.Form;
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/jobdomain/control/JobDomainPeasSessionController.java]: MINOR Remove this unused import 'org.silverpeas.core.contribution.content.form.Form'. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core-war/src/main/java/org/silverpeas/web/directory/control/DirectorySessionController.java [hunk 12]
```diff
@@ -50,10 +61,12 @@
 import org.silverpeas.core.socialnetwork.relationship.RelationShipService;
 import org.silverpeas.core.template.SilverpeasTemplate;
 import org.silverpeas.core.template.SilverpeasTemplateFactory;
+import org.silverpeas.core.util.CollectionUtil;
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/directory/control/DirectorySessionController.java]: MINOR Remove this unused import 'org.silverpeas.core.util.CollectionUtil'. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core-war/src/main/java/org/silverpeas/web/directory/control/DirectorySessionController.java [hunk 13]
```diff
@@ -799,4 +810,91 @@ private DirectoryItem getDirectoryItem(final DirectoryItemList allUsers,
...
+    if (template != null) {
+      // build the xmlSubQuery according to the dataRecord object
+      String templateFileName = template.getFileName();
+      String templateName = templateFileName.substring(0, templateFileName.lastIndexOf("."));
+      if (xmlData != null) {
+        for (String fieldName : xmlData.getFieldNames()) {
+          try {
+            Field field = xmlData.getField(fieldName);
+            String fieldValue = field.getStringValue();
+            if (fieldValue != null && fieldValue.trim().length() > 0) {
```
Comment (silveruser) [inline core-war/src/main/java/org/silverpeas/web/directory/control/DirectorySessionController.java]: MINOR Refactor this code to not nest more than 4 if/for/while/switch/try statements. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS134)

**Comment by silveruser** [comment]
SonarQube analysis reported 17 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 13 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [UserIndexation.java#L35](https://github.com/Silverpeas/Silverpeas-Core/blob/4de90276cd018a7d4027890a5e32b74e3a3d6530/core-library/src/main/java/org/silverpeas/core/admin/user/UserIndexation.java#L35): Remove this unused import 'org.silverpeas.core.index.search.model.MatchingIndexEntry'. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [UserIndexation.java#L36](https://github.com/Silverpeas/Silverpeas-Core/blob/4de90276cd018a7d4027890a5e32b74e3a3d6530/core-library/src/main/java/org/silverpeas/core/admin/user/UserIndexation.java#L36): Remove this unused import 'org.silverpeas.core.index.search.model.ParseException'. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [UserIndexation.java#L37](https://github.com/Silverpeas/Silverpeas-Core/blob/4de90276cd018a7d4027890a5e32b74e3a3d6530/core-library/src/main/java/org/silverpeas/core/admin/user/UserIndexation.java#L37): Remove this unused import 'org.silverpeas.core.index.search.model.QueryDescription'. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [UserIndexation.java#L41](https://github.com/Silverpeas/Silverpeas-Core/blob/4de90276cd018a7d4027890a5e32b74e3a3d6530/core-library/src/main/java/org/silverpeas/core/admin/user/UserIndexation.java#L41): Remove this unused import 'java.util.ArrayList'. [rule](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by mmoqui** [comment]
This PR was merged with some changes done by the integrator.

**Pull request closed without merging**